htlcswitch/payment_result.go

```diff
@@ -104,3 +114,155 @@ func deserializePaymentResult(r io.Reader) (*PaymentResult, error) {
 
 	return p, nil
 }
+
+// paymentResultStore is a persistent store that stores any results of HTLCs in
+// flight on the network. Since payment results are inherently asynchronous, it
+// is used as a common access point for senders of HTLCs, to know when a result
+// is back. The Switch will checkpoint any and any received result to the
+// store, and the store will keep results and notify the callers about them.
+type paymentResultStore struct {
+	db *channeldb.DB
+
+	// results is a map from paymentIDs to channels where subscribers to
+	// payment results will be notified.
+	results    map[uint64][]chan *PaymentResult
+	resultsMtx sync.Mutex
+}
+
+func newPaymentResultStore(db *channeldb.DB) *paymentResultStore {
+	return &paymentResultStore{
+		db:      db,
+		results: make(map[uint64][]chan *PaymentResult),
+	}
+}
+
+// storePaymentResult stores the PaymentResult for the given paymentID, and
+// notifies any subscribers.
+func (store *paymentResultStore) storePaymentResult(paymentID uint64,
+	result *PaymentResult) error {
+
+	// Serialize the payment result.
+	var b bytes.Buffer
+	if err := serializePaymentResult(&b, result); err != nil {
+		return err
+	}
+
+	var paymentIDBytes [8]byte
+	binary.BigEndian.PutUint64(paymentIDBytes[:], paymentID)
+
+	err := store.db.Batch(func(tx *bbolt.Tx) error {
+		paymentResults, err := tx.CreateBucketIfNotExists(
+			paymentStoreBucketKey,
+		)
+		if err != nil {
+			return err
+		}
+
+		return paymentResults.Put(paymentIDBytes[:], b.Bytes())
+	})
+	if err != nil {
+		return err
+	}
+
+	// Now that the result is stored in the database, we can notify any
```

Comment:
If the subscribe happens at this point, there will be a duplicate notification?

Reply:
Fixed by using multimutex on the payment ID.